pkg/readiness/objset.go

```diff
@@ -27,9 +27,35 @@ type objKey struct {
 	namespacedName types.NamespacedName
 }
 
-type objSet map[objKey]struct{}
-
 func (k objKey) String() string {
 	return fmt.Sprintf("%s [%s]", k.namespacedName.String(), k.gvk.String())
+}
+
+type objData struct {
+	retries int
+}
 
+// decrementRetries handles objData retries, and returns `true` if it's time to delete the objData entry
+func (o *objData) decrementRetries() bool {
+	// if retries is less than 0, allowed retries are infinite
+	if o.retries < 0 {
+		return false
+	}
+
+	// If we have retries left, use one
+	if o.retries > 0 {
+		o.retries--
+		return false
+	}
+
+	// if we have zero retries, we can delete
+	return true
 }
+
+type objDataFactory func() *objData
```

Comment:
nit: from what I can tell, all uses of these factory functions always immediately dereference the pointer. Should we just return a value?

Reply:
It's dereferences across the board.  Personally, I vote in favor of returning `objData` by value, as that's consistent with the usage in the package (values of an `objSet` are `objData` structs, not pointers to said structs).  

I'm going to switch this to returning a value unless @maxsmythe feels strongly.

Reply:
+1